openmc/universe.py

```diff
@@ -519,3 +519,53 @@ def create_xml_subelement(self, xml_element):
                 # Append the Universe ID to the subelement and add to Element
                 cell_element.set("universe", str(self._id))
                 xml_element.append(cell_element)
+
+    def _count_cell_instances(self):
+        """Count the number of instances for each cell in the universe, and
+        record the count in the :attr:`Cell.num_instances` properties."""
+
+        for cell in self.cells.values():
+            # Increment the number of cell instances
+            cell._num_instances += 1
+
+            # If universe-filled, recursively count cells in filling universe
+            if cell.fill_type == 'universe':
+                cell.fill._count_cell_instances()
+
+            # If lattice-filled, recursively call for all universes in lattice
+            elif cell.fill_type == 'lattice':
+                latt = cell.fill
+
+                # Count instances in each universe in the lattice
+                for index in latt.indices:
+                    if latt.ndim == 3:
+                        univ = latt.universes[index[0]][index[1]][index[2]]
+                    else:
+                        univ = latt.universes[index[0]][index[1]]
+                    univ._count_cell_instances()
+
+    def _count_material_instances(self):
+        """Count the number of instances for each material in the Universe, and
+        record the count in the :attr:`Material.num_instances` properties."""
+
+        for cell in self.cells.values():
+
+            # If material-filled, we are finished with all levels
+            if cell.fill_type == 'material':
+                cell.fill._num_instances += 1
```

Comment:
Thinking about this a little more, it is a little complicated. Each time you recurse to the lowest level, you would only want to increment the instance count for the corresponding material, but there's no way to know how many times you've reached this cell (and hence which material in `cell.fill` you should be incrementing), that is unless you count cell and material instances in the same method, in which case you could do something like
```Python
elif cell.fill_type == 'distribmat':
    cell.fill[cell._num_instances]._num_instances += 1
```